--- crypto/evp/evp_orig.h	2015-10-07 20:11:14.000000000 +0300
+++ crypto/evp/evp.h	2015-10-07 20:11:56.000000000 +0300
@@ -422,6 +422,18 @@
 # define         EVP_CTRL_TLS1_1_MULTIBLOCK_ENCRYPT      0x1a
 # define         EVP_CTRL_TLS1_1_MULTIBLOCK_DECRYPT      0x1b
 # define         EVP_CTRL_TLS1_1_MULTIBLOCK_MAX_BUFSIZE  0x1c
+/* EVP_CTRL_SET_SBOX takes the char* specifying S-boxes */
+# define         EVP_CTRL_SET_SBOX                       0x1d
+/* 
+ * EVP_CTRL_SBOX_USED takes a 'size_t' and 'char *', pointing at a
+ * pre-allocated buffer with specified size 
+ */
+# define         EVP_CTRL_SBOX_USED                      0x1e
+/* EVP_CTRL_KEY_MESH takes 'size_t' number of bytes to mesh the key after, 0 switches meshing off */
+# define         EVP_CTRL_KEY_MESH                       0x1f
+/* EVP_CTRL_BLOCK_PADDING_MODE takes the char* specifying padding mode */
+# define         EVP_CTRL_BLOCK_PADDING_MODE             0x20
+
 
 /* RFC 5246 defines additional data to be 13 bytes in length */
 # define         EVP_AEAD_TLS1_AAD_LEN           13
